plugins/intrinsics/src/main/java/org/qbicc/plugin/intrinsics/core/CoreIntrinsics.java

```diff
@@ -368,12 +370,106 @@ public static void registerJavaLangThreadIntrinsics(CompilationContext ctxt) {
         Intrinsics intrinsics = Intrinsics.get(ctxt);
         ClassContext classContext = ctxt.getBootstrapClassContext();
 
+        final int threadAlive = 0x0001;
+        final int threadTerminated = 0x0002;
+        final int threadRunnable = 0x0004;
+
         ClassTypeDescriptor jltDesc = ClassTypeDescriptor.synthesize(classContext, "java/lang/Thread");
+        ClassTypeDescriptor vmHelpersDesc = ClassTypeDescriptor.synthesize(classContext, "org/qbicc/runtime/main/VMHelpers");
+        ClassTypeDescriptor pthreadPtrDesc = ClassTypeDescriptor.synthesize(classContext, "org/qbicc/runtime/posix/PThread$pthread_t_ptr");
+        ClassTypeDescriptor voidPtrDesc = ClassTypeDescriptor.synthesize(classContext, "org/qbicc/runtime/CNative$void_ptr");
+        ClassTypeDescriptor voidUnaryfunctionPtrDesc = ClassTypeDescriptor.synthesize(classContext, "org/qbicc/runtime/CNative$void_ptr_unaryoperator_function_ptr");
+
         MethodDescriptor returnJlt = MethodDescriptor.synthesize(classContext, jltDesc, List.of());
+        MethodDescriptor voidDesc = MethodDescriptor.synthesize(classContext, BaseTypeDescriptor.V, List.of());
+        MethodDescriptor voidIntDesc = MethodDescriptor.synthesize(classContext, BaseTypeDescriptor.V, List.of(BaseTypeDescriptor.I));
+        MethodDescriptor booleanDesc = MethodDescriptor.synthesize(classContext, BaseTypeDescriptor.Z, List.of());
 
-        StaticIntrinsic currentThread = (builder, target, arguments) -> builder.currentThread();
+        StaticIntrinsic nopStatic = (builder, target, arguments) -> ctxt.getLiteralFactory().zeroInitializerLiteralOfType(target.getType().getReturnType());
+        InstanceIntrinsic nopInstance = (builder, instance, target, arguments) -> ctxt.getLiteralFactory().zeroInitializerLiteralOfType(target.getType().getReturnType());
+
+        /* registerNatives */
+        intrinsics.registerIntrinsic(jltDesc, "registerNatives", voidDesc, nopStatic);
 
+        /* public static native Thread currentThread(); */
+        StaticIntrinsic currentThread = (builder, target, arguments) -> builder.currentThread();
         intrinsics.registerIntrinsic(jltDesc, "currentThread", returnJlt, currentThread);
+
+        /* VMHelpers.java - threadWrapper: helper method for java.lang.Thread.start0 */
+        MethodDescriptor threadWrapperNativeDesc = MethodDescriptor.synthesize(classContext, voidPtrDesc, List.of(voidPtrDesc));
+        StaticIntrinsic threadWrapperNative = (builder, target, arguments) -> {
+            Value threadVoidPtr = arguments.get(0);
+
+            DefinedTypeDefinition jlt = classContext.findDefinedType("java/lang/Thread");
+            LoadedTypeDefinition jltVal = jlt.load();
+            ValueType jltType = jltVal.getType().getReference();
+            Value threadObject = builder.bitCast(threadVoidPtr, (WordType)jltType);
```

Comment:
This makes sense as a starting point but will be unsafe once we have a GC.  We can't pass object pointers through native functions without having either pinned them or used some level of indirection (ie: a jni handle) so the GC can update the real pointer.

It looks like you've already started down this path by adding `saveNativeThread()`.  In J9, we kept a circular linked list of J9VMThread pointers in the VM structure that the GC could use to find all threads to ensure they were coalesced when updating the heap.

We'll need a similar list and the list elements can do double duty - act as a list of roots for the GC and provide the indirection here so we don't lose the thread object

Reply:
That is the plan for `saveNativeThread`, to add that information to a global list. That will be a future pull request.

Reply:
And to not pass the "naked" object pointer into this function?  Then I'm OK with the approach

Reply:
Ah yes and pass in the list element. Sorry I think I was getting confused about the meaning of "native functions" for this. You mean don't pass direct object pointers to things like c libraries & exported functions since they will not be covered by the JNI level of indirection?

Reply:
> You mean don't pass direct object pointers to c libraries (pthread_create) since they will not be covered by the JNI level of indirection?

Exactly.  The GC loses track of the direct object pointers once they go off into c libraries.  So we need something - the list element or a JNI handle - to allow the GC to track and update the object pointer regardless of where it may end up in native (c) code.

Reply:
Got it! Yes I will update the threadWrapper parameter to be the list element instead when that exists.